### DISCUSSION TITLE
Error/Warning by following the pyMor balanced Truncation Tutorial

### DISCUSSION THREAD

**Discussion by NoraYV** [pymor/pymor#1591] [Q&A] [answered]
I tried to follow the _Documentation Tutorial: Reducing an LTI system using balanced truncation_ step by step. At the last step before calculating the _relative errors in different system norms_  I get the warning message `UserWarning: Data has no positive values, and therefore cannot be log-scaled.
  out = ax.loglog(freq, mag, **mpl_kwargs)`. I don´t know what I can do to prevent this warning. 

 Also in the last step to calculating the _relative errors in different system norms_ while calculating the _Relative Hinf error_ I get the Error `Traceback (most recent call last):
  File "C:\Users\Yazdandoost\PycharmProjects\pythonProject\Tutorial_BT.py", line 84, in <module>
    print(f"Relative Hinf error:   {err.hinf_norm() / fom.hinf_norm():.3e}")
  File "C:\Users\Yazdandoost\Anaconda3\envs\GitHub\lib\site-packages\pymor\core\cache.py", line 379, in wrapper
    return function(self, *args, **kwargs)
  File "C:\Users\Yazdandoost\Anaconda3\envs\GitHub\lib\site-packages\pymor\models\iosys.py", line 710, in hinf_norm
    return self.linf_norm(mu=mu, return_fpeak=return_fpeak, ab13dd_equilibrate=ab13dd_equilibrate)
  File "C:\Users\Yazdandoost\Anaconda3\envs\GitHub\lib\site-packages\pymor\core\cache.py", line 379, in wrapper
    return function(self, *args, **kwargs)
  File "C:\Users\Yazdandoost\Anaconda3\envs\GitHub\lib\site-packages\pymor\models\iosys.py", line 821, in linf_norm
    raise NotImplementedError
NotImplementedError`
How can I avoid this error? I followed the tutorial, so why doesn´t I get a result?

**Comment by pmli** [comment, marked as answer]
Hi @NoraYV, thank you for the report.

I'm not sure what is the cause of the first error, but it sounds like `err = fom - rom` was replaced with `err = fom - fom` (i.e., substracting the FOM from itself). Otherwise, it would be helpful for me to see exactly which code was used.

For the second error, the most likely cause is that `slycot` is not installed. It looks like you're using Anaconda, so `conda install -c conda-forge slycot` should work.

**Reply by NoraYV** [reply]
Hey @pmli thank you for your support.

Here is my full code. I marked the err = fom-rom fat. 
`import matplotlib.pyplot as plt
import numpy as np
import scipy.sparse as sps
from pymor.models.iosys import LTIModel
from pymor.reductors.bt import BTReductor

plt.rcParams['axes.grid'] = True

k = 50
n = 2*k+1
E = sps.eye(n, format='lil') # Einheitspatrix
E[0,0] = E[-1, -1] = 0.5 # Erster und letzter Wert werden zu 0,5 gesetzt
E = E.tocsc()

d0 = n*[-2*(n-1)**2]
d1 = (n - 1) * [(n-1)**2]
A = sps.diags([d1, d0, d1], [-1, 0, 1], format='lil')
A[0,0] = A[-1, -1] = -n*(n-1)
A = A.tocsc()

B = np.zeros((n, 2))
B[:, 0] = 1
B[0,0] = B[-1, 0] = 0.5
B[0,1] = n-1

C = np.zeros((3, n))
C[0,0] = C[1, k] = C[2, -1] = 1

fom = LTIModel.from_matrices(A, B, C, E=E)

'''Annahme zur Balanced Truncation:
System erster Ordnung: Ex(abgeleitet)=Ax+bu; y=Cx+Du
andernfalls kann x auch durch Tx(schlange/Nährung) oder andere Multiplikationen genutzt werden.
Weitere Infos: https://docs.pymor.org/2021-2-0/tutorial_bt.html
Wichtig ist dass eine invertierbate Transformationsmatrix T und S der Ordnung nxn exestieren, sodass Nährungen wie folgt entstehen:
E(schlange)=S(transformiert)ET=I(Einheitsmatrix)
A(Schlange)=S(transformiert)AT
B(Schlange)=S(transformiert)B
C(Schlange)=CT
sodass die Gramschen Matrizen P(schlange) und Q(schlange) identisch sind und aus aus den diagonalen Hankel singular Werten entstehen. 
Tutorial für die Gramschen Matrizen: https://docs.pymor.org/2021-2-0/tutorial_lti_systems.html
bzw. als Tutorial_Gramsche_Matrizen abgespeichert'''

#System BR reduzieren
bt = BTReductor(fom)

error_bounds = bt.error_bounds()
hsv = fom.hsv()
fig, ax = plt.subplots()
ax.semilogy(range(1, len(error_bounds)+1), error_bounds, '.-')
ax.semilogy(range(1, len(hsv)), hsv[1:], '.-')
ax.set_xlabel('Reduced order')
#_ = ax.set_title(r'Upper an lower $\mathcal{H}_\infty$ error bounds')
#plt.show()

'''Modell auf eine Gewisse Ordnung reduzieren
Beispiel der Ordnung 10'''
rom = bt.reduce(10)

'''Vergleich der Amplituden und Bode Diagramme des vollständigen und des reduzierten Modells'''
#Amplituden vergleichen
w = np.logspace(-2, 8, 300)
fig, ax = plt.subplots()
fom.transfer_function.mag_plot(w, ax=ax, label='FOM')
rom.transfer_function.mag_plot(w, ax=ax, linestyle='--', label='ROM')
#_ = ax.legend()
#plt.show()
fig, axs = plt.subplots(6, 2, figsize=(12, 24), sharex=True, constrained_layout=True)
fom.transfer_function.bode_plot(w, ax=axs)
#_ = rom.transfer_function.bode_plot(w, ax=axs, linestyle='--')
#plt.show()

'''Fehler beider Systeme darstellen'''
**err = fom - rom**
#Amplituendenfehler
_ = err.transfer_function.mag_plot(w)
#Bode Fehler
_ = err.transfer_function.bode_plot(w)
plt.show()

'''relative Fehler bestimmen'''
#print(f"Relative Hinf error:   {err.hinf_norm() / fom.hinf_norm():.3e}")
#print(f'Relative H2 error:     {err.h2_norm() / fom.h2_norm():.3e}')
#print(f'Relative Hankel error: {err.hankel_norm() / fom.hankel_norm():.3e}')`


From this Code I get the warning `UserWarning: Data has no positive values, and therefore cannot be log-scaled. out = ax.loglog(freq, mag, **mpl_kwargs)`


I tried to install `slycot` as you say, but get this error and done know how to solve it.
`## Package Plan ##

  environment location: C:\Users\Yazdandoost\Anaconda3

  added / updated specs:
    - slycot


The following NEW packages will be INSTALLED:

  _anaconda_depends  pkgs/main/win-64::_anaconda_depends-2021.11-py39_0
  astroid            pkgs/main/win-64::astroid-2.6.6-py39haa95532_0
  setuptools         conda-forge/win-64::setuptools-60.9.3-py39hcbf5309_0

The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates    pkgs/main::ca-certificates-2021.10.26~ --> conda-forge::ca-certificates-2021.10.8-h5b45459_0
  openssl              pkgs/main::openssl-1.1.1l-h2bbff1b_0 --> conda-forge::openssl-1.1.1l-h8ffe710_0

The following packages will be DOWNGRADED:

  anaconda                                   2021.11-py39_0 --> custom-py39_1


Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: failed

RemoveError: 'setuptools' is a dependency of conda and cannot be removed from
conda's operating environment.`

**Reply by pmli** [reply]
Interesting. It seems the issue appears when `matplotlib` tries to use a single figure to plot everything. Uncommenting all `plt.show()` solves the issue for me. (An alternative is creating figures "manually"; e.g., `fig, ax = plt.subplots(); err.transfer_function.mag_plot(w, ax=ax)`.)

Concerning the installation of `slycot`, it seems to me that the issue comes from using the base environment. I would suggest creating a new [`conda` environment](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html), installing the packages there, and then running the script from there. The following should work.

```bash
conda create -n myenv python=3.7
conda activate myenv
conda install numpy scipy matplotlib
conda install -c conda-forge pymor slycot
python script_name.py
```

**Reply by NoraYV** [reply]
Thank you so much! After uncommenting als `plt.show()` the code runs completely and give the same result for each error as the tutorial!

**Reply by pmli** [reply]
Great. Did the installation via a new environment work or did you use a different method?
